src/server/raft_client/fresh.rs

```diff
@@ -610,25 +658,216 @@ where
             // need to fallback to use legacy API.
-            res = back_end.call(&client, addr.clone()).await;
+            let f = back_end.call(&client, addr.clone());
+            res = f.await;
         }
         match res {
             Ok(()) => {
                 error!("connection fail"; "store_id" => back_end.store_id, "addr" => addr, "error" => "require fallback even with legacy API");
             }
             Err(_) => {
                 error!("connection abort"; "store_id" => back_end.store_id, "addr" => addr);
-                back_end.router.broadcast_unreachable(back_end.store_id);
+                back_end
+                    .builder
+                    .router
+                    .broadcast_unreachable(back_end.store_id);
+            }
+        }
+    }
+}
+
+/// A global connection pool.
+///
+/// All valid connections should be stored as a record. Once it's removed
+/// from the struct, all cache clone should also remove it at some time.
+#[derive(Default)]
+struct ConnectionPool {
+    connections: HashMap<(u64, usize), Arc<Queue>>,
+    tombstone_stores: HashSet<u64>,
+}
+
+/// Queue in cache.
+struct CachedQueue {
+    queue: Arc<Queue>,
+    /// If a msg is enqueued, but the queue has not been notified for polling,
+    /// it will be marked to true. And all dirty queues are expected to be
+    /// notified during flushing.
+    dirty: bool,
+}
+
+/// A raft client that can manages connections correctly.
+///
+/// A correct usage of raft client is:
+///
+/// ```text
+/// for m in msgs {
+///     if !raft_client.send(m) {
+///         // handle error.   
+///     }
+/// }
+/// raft_client.flush();
+/// ```
+pub struct RaftClient<S, R> {
+    pool: Arc<Mutex<ConnectionPool>>,
+    cache: LruCache<(u64, usize), CachedQueue>,
+    need_flush: Vec<(u64, usize)>,
+    future_pool: Arc<ThreadPool>,
+    builder: ConnectionBuilder<S, R>,
+}
+
+impl<S, R> RaftClient<S, R>
+where
+    S: StoreAddrResolver + Send + 'static,
+    R: RaftStoreRouter<RocksEngine> + Send + 'static,
+{
+    #[allow(unused)]
+    pub fn new(builder: ConnectionBuilder<S, R>) -> RaftClient<S, R> {
+        let future_pool = Arc::new(
+            yatp::Builder::new(thd_name!("raft-stream"))
+                .max_thread_count(1)
+                .build_future_pool(),
+        );
+        RaftClient {
+            pool: Arc::default(),
+            cache: LruCache::with_capacity_and_sample(0, 7),
+            need_flush: vec![],
+            future_pool,
+            builder,
+        }
+    }
+
+    /// Loads connection from pool.
+    ///
+    /// Creates it if it doesn't exist. `false` is returned if such connection
+    /// can't be established.
+    fn load_stream(&mut self, store_id: u64, conn_id: usize) -> bool {
+        let (s, pool_len) = {
+            let mut pool = self.pool.lock().unwrap();
+            if pool.tombstone_stores.contains(&store_id) {
+                return false;
```

Comment:
No, if dns resolve fails, it will retry automatically. Only when a store becomes tombstone will be push into `tombstone_stores`. When a store becomes tombstone, it can't be up again.